### ISSUE TITLE
There is a flaw in the handling of custom backends

### ISSUE THREAD

**Issue by jwiegley** [libgit2/libgit2#1146]
Higher priority custom backends, if unable to perform a certain action, are supposed to automatically fall back to a lower priority backend.  The logic in `git_odb_write`, for example, indicates this.

However, `git_blob_create_frombuffer` does not call `git_odb_write`, but only `git_odb_open_wstream`, which -- if stream writing is not supported -- creates a fake stream using _the highest priority backend that supports writestream or write_.

The flaw here is that if the write method backing the fake stream fails (expecting an automatic fallback), no fallback occurs because there is no logic to create a new fake stream using the next most applicable backend.

In short, the above makes it impossible to create a tracing backend for writes.  If you try, you end up not being able to write objects.

Since I need this kind of intelligent fallback for regular writes I will fix the Blob behavior and submit a pull request.  I wanted to track the issue here to get comments and to see if anyone has already solved this.
